Lib/functools.py

```diff
@@ -653,6 +653,48 @@ def cache(user_function, /):
     return lru_cache(maxsize=None)(user_function)
 
 
+################################################################################
+### Function variant registry
+################################################################################
+
+# {key: (max lineno, [variant])}
+_variant_registry = {}
+
+
+def register_variant(key, variant):
+    """Register a function variant."""
+    _variant_registry.setdefault(key, []).append(variant)
+
+
+def get_variants(key):
+    """Get all function variants for the given key."""
+    return _variant_registry.get(key, [])
+
+
+def clear_variants(key=None):
+    """Clear all variants for the given key (or all keys)."""
+    if key is None:
+        _variant_registry.clear()
+    else:
+        _variant_registry.pop(key, None)
+
+
+def get_key_for_callable(func):
```

Comment:
Why is this exposed as a public function, rather than being a private implementation detail? I'm struggling to think of a situation when a user would want to call it, and I feel like it limits our flexibility to change the implementation of the variants registry later on, if we want to. I'd rename it as `_get_key_for_callable`, and remove the docs.

Reply:
My starting point here was that we need to preserve backward compatibility on `@overload`, so we should not fail if people call `@overload` on some exotic object that `get_key_for_callable` doesn't know what to do with. However, we can achieve that same goal without making `get_key_for_callable` public, so I'll change it to private as you suggest.